domain/info.go

```diff
@@ -0,0 +1,180 @@
+	info           *ServerInfo
+	serverInfoPath string
+}
+
+// ServerInfo is server static information.
+// It will not update when server running. So please only put static information in ServerInfo struct.
+type ServerInfo struct {
+	ServerVersionInfo
+	ID         string `json:"ddl_id"`
+	IP         string `json:"ip"`
+	Port       uint   `json:"listening_port"`
+	StatusPort uint   `json:"status_port"`
+	Lease      string `json:"lease"`
+}
+
+// ServerVersionInfo is the server version and git_hash.
+type ServerVersionInfo struct {
+	Version string `json:"version"`
+	GitHash string `json:"git_hash"`
+}
+
+// NewInfoSyncer return new InfoSyncer. It is exported for testing.
+func NewInfoSyncer(id string, etcdCli *clientv3.Client) *InfoSyncer {
+	return &InfoSyncer{
+		etcdCli:        etcdCli,
+		info:           getServerInfo(id),
+		serverInfoPath: fmt.Sprintf("%s/%s", ServerInformationPath, id),
+	}
+}
+
+func getServerInfo(id string) *ServerInfo {
+	cfg := config.GetGlobalConfig()
+	info := &ServerInfo{
+		ID:         id,
+		IP:         cfg.AdvertiseAddress,
+		Port:       cfg.Port,
+		StatusPort: cfg.Status.StatusPort,
+		Lease:      cfg.Lease,
+	}
+	info.Version = mysql.ServerVersion
+	info.GitHash = printer.TiDBGitHash
+	return info
+}
+
+//GetServerInfo gets self server static information.
+func (is *InfoSyncer) GetServerInfo() *ServerInfo {
+	return is.info
+}
+
+// GetServerInfoByID gets server static information from Etcd.
+func (is *InfoSyncer) GetServerInfoByID(ctx context.Context, id string) (*ServerInfo, error) {
+	if is.etcdCli == nil || id == is.info.ID {
+		return is.info, nil
+	}
+	key := fmt.Sprintf("%s/%s", ServerInformationPath, id)
+	infoMap, err := getInfo(ctx, is.etcdCli, key, keyOpDefaultRetryCnt, keyOpDefaultTimeout)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info, ok := infoMap[id]
+	if !ok {
+		return nil, errors.Errorf("[infoSyncer] get %s failed", key)
+	}
+	return info, nil
+}
+
+// GetAllServerInfo gets all servers static information from Etcd.
+func (is *InfoSyncer) GetAllServerInfo(ctx context.Context) (map[string]*ServerInfo, error) {
+	allInfo := make(map[string]*ServerInfo)
+	if is.etcdCli == nil {
+		allInfo[is.info.ID] = is.info
+		return allInfo, nil
+	}
+	allInfo, err := getInfo(ctx, is.etcdCli, ServerInformationPath, keyOpDefaultRetryCnt, keyOpDefaultTimeout, clientv3.WithPrefix())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return allInfo, nil
+}
+
+// StoreServerInfo stores self server static information to Etcd when domain Init.
+func (is *InfoSyncer) StoreServerInfo(ctx context.Context) error {
+	if is.etcdCli == nil {
+		return nil
+	}
+	infoBuf, err := json.Marshal(is.info)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = ddl.PutKVToEtcd(ctx, is.etcdCli, keyOpDefaultRetryCnt, is.serverInfoPath, hack.String(infoBuf))
+	return errors.Trace(err)
+}
+
+// RemoveServerInfo remove self server static information from Etcd.
+func (is *InfoSyncer) RemoveServerInfo() {
```

Comment:
If TiDB have no chance to call `RemoveServerInfo` because panic or crashed, how to deal with the serverInfo key?

Reply:
I guess there is a lifetime for the key-value. Like the ddl schema version info in the etcd.

Reply:
I will add lease for the key.